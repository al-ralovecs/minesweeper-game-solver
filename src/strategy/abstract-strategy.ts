import BoardDto from "../dto/board.dto";
import CoordinateDto from "../dto/coordinate.dto";

export default abstract class AbstractStrategy
{
    protected readonly board: BoardDto;
    protected totalMinesCount: number;
    protected solution: CoordinateDto;
    protected isHasSolution: boolean = false;

    protected constructor(board: BoardDto, totalMinesCount: number)
    {
        this.board = board;
        this.totalMinesCount = totalMinesCount;
    }

    public abstract apply();

    public get hasSolution(): boolean
    {
        return this.isHasSolution && typeof this.solution !== 'undefined';
    }

    public get getNextMove(): CoordinateDto
    {
        if (typeof this.solution === 'undefined') {
            throw Error('[Strategy] Fail on attempt to query a solution that does not exist');
        }

        return this.solution;
    }
}
