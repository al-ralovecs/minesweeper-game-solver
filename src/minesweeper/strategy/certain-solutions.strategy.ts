import { AbstractStrategy, StrategyType } from './abstract-strategy';
import BoardStateDto from '../dto/board-state.dto';
import WitnessWebDto from '../dto/witness-web.dto';
import ProbabilityDistributionDto from '../dto/probability-distribution.dto';
import ActionDto, { ActionType } from '../dto/action.dto';

export default class CertainSolutionsStrategy extends AbstractStrategy {
	private readonly wholeEdge: WitnessWebDto;
	private readonly probabilityDistribution: ProbabilityDistributionDto;

	public constructor(
		boardState: BoardStateDto,
		wholeEdge: WitnessWebDto,
		probabilityDistribution: ProbabilityDistributionDto,
	) {
		super(boardState);

		this.wholeEdge = wholeEdge;
		this.probabilityDistribution = probabilityDistribution;
	}

	// there is only one solution
	// or the solutions are certainties
	protected get isStrategyApplicable(): boolean {
		return 1 === this.probabilityDistribution.bestCandidates.length
			|| this.probabilityDistribution.foundCertainty;
	}

	protected applyStrategy(): void {
		this.probabilityDistribution.bestCandidates.forEach((cl) => {
			this.boardState.setAction = cl.buildAction(StrategyType.CertainSolutions);
		});
		
		if (this.probabilityDistribution.foundCertainty) {
			this.probabilityDistribution.mines.forEach((loc) => {
				this.boardState.setAction = new ActionDto(loc, ActionType.Flag, StrategyType.CertainSolutions, 1);
			});
		}
	}

	protected get getMoveMethod(): StrategyType {
		return StrategyType.CertainSolutions;
	}
}
