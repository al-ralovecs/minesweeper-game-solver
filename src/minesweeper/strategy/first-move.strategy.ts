import { AbstractStrategy, StrategyType } from './abstract-strategy';
import LocationDto from '../dto/location.dto';
import ActionDto, { ActionType } from '../dto/action.dto';

export default class FirstMoveStrategy extends AbstractStrategy {
    protected get isStrategyApplicable(): boolean {
        return this.boardState.getTotalUnrevealedCount === this.boardState.height * this.boardState.width;
    }

    protected applyStrategy(): void {
        const y: number = Math.round(this.boardState.height / 2) - 1;
        const x: number = Math.round(this.boardState.width / 2) - 1;

        if (0 > y || 0 > x || this.boardState.height <= y || this.boardState.width <= x) {
            throw Error(`[FirstMoveStrategy] Proposed first move (${x}, ${y}) is out of board bounds`);
        }

        this.boardState.setAction = new ActionDto(
            new LocationDto(y, x),
            ActionType.Clear,
            StrategyType.FirstMove,
            1,
        );
    }

    protected get getMoveMethod(): StrategyType {
        return StrategyType.FirstMove;
    }
}
