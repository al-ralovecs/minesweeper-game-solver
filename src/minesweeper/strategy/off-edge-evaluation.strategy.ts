import { AbstractStrategy, StrategyType } from './abstract-strategy';

import BoardStateDto from '../dto/board-state.dto';
import ProbabilityDistributionDto from '../dto/probability-distribution.dto';
import EvaluateLocationsService from '../service/evaluate-locations.service';

const OFF_EDGE_TOLERANCE: number = 0.97;

export default class OffEdgeEvaluationStrategy extends AbstractStrategy {
    private readonly probabilityDistribution: ProbabilityDistributionDto;
    private readonly evaluateLocationsService: EvaluateLocationsService;

    public constructor(
        boardState: BoardStateDto,
        probabilityDistribution: ProbabilityDistributionDto,
        evaluateLocationsService: EvaluateLocationsService,
    ) {
        super(boardState);

        this.probabilityDistribution = probabilityDistribution;
        this.evaluateLocationsService = evaluateLocationsService;
    }

    protected get isStrategyApplicable(): boolean {
        // are clears off the edge within the permitted cut-off ?
        return 0 < this.probabilityDistribution.offEdgeProbability - this.probabilityDistribution.bestProbability * OFF_EDGE_TOLERANCE
            && ! this.probabilityDistribution.foundCertainty;
    }

    protected applyStrategy(): void {
        this.evaluateLocationsService.addOffEdgeCandidates(this.boardState.getAllUnrevealedSquares);
        this.evaluateLocationsService.evaluateLocations(this.probabilityDistribution.bestCandidates);

        if (! this.evaluateLocationsService.hasBestMove) {
            return;
        }

        this.boardState.setAction = this.evaluateLocationsService.getBestMove(this.getMoveMethod);
    }

    protected get getMoveMethod(): StrategyType {
        return StrategyType.OffEdgeEvaluation;
    }
}
