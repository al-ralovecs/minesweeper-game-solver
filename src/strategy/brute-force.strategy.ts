import { AbstractStrategy, StrategyType } from './abstract-strategy';
import BoardStateDto from "../dto/board-state.dto";
import WitnessWebDto from "../dto/witness-web.dto";
import ProbabilityDistributionDto from "../dto/probability-distribution.dto";

const preferences = {
    ALLOWABLE_SOLUTIONS_MAXIMUM: 4000,
};

export default class BruteForceStrategy extends AbstractStrategy
{
    private readonly wholeEdge: WitnessWebDto;
    private readonly probabilityDistribution: ProbabilityDistributionDto;

    public constructor(
        boardState: BoardStateDto,
        wholeEdge: WitnessWebDto,
        probabilityDistribution: ProbabilityDistributionDto
    ) {
        super(boardState);

        this.wholeEdge = wholeEdge;
        this.probabilityDistribution = probabilityDistribution;
    }

    // if the number of candidate solutions is not greater than the allowable maximum
    protected get isStrategyApplicable(): boolean
    {
        return false;

        // return 0 > this.probabilityDistribution.finalSolutionsCount - preferences.ALLOWABLE_SOLUTIONS_MAXIMUM
        //     && ! this.probabilityDistribution.foundCertainty;
    }

    protected applyStrategy(): void
    {
        // @todo: implement
    }

    protected get getMoveMethod(): StrategyType
    {
        return StrategyType.BruteForce;
    }
}
