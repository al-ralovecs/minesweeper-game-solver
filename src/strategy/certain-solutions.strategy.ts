import {AbstractStrategy, StrategyType} from "./abstract-strategy";
import WitnessWebDto from "../dto/witness-web.dto";
import ProbabilityDistributionDto, { PROBABILITY_ENGINE_TOLERANCE } from "../dto/probability-distribution.dto";
import BoardStateDto from "../dto/board-state.dto";
import ProbabilityEngineService from "../service/probability-engine.service";

export default class CertainSolutionsStrategy extends AbstractStrategy
{
	private readonly wholeEdge: WitnessWebDto;
	private readonly probabilityDistribution: ProbabilityDistributionDto;

	public constructor(
		boardState: BoardStateDto,
		wholeEdge: WitnessWebDto,
		probabilityDistribution: ProbabilityDistributionDto
	) {
		super(boardState);

		this.wholeEdge = wholeEdge;
		this.probabilityDistribution = probabilityDistribution;
	}

	// there is only one solution
	// or the solutions are certainties
	protected get isStrategyApplicable(): boolean
	{
		return 1 === ProbabilityEngineService.getBestCandidates(
				this.boardState,
				this.probabilityDistribution,
				PROBABILITY_ENGINE_TOLERANCE
			).length
			|| this.probabilityDistribution.foundCertainty;
	}

	protected applyStrategy(): void {
	}

	protected get getMoveMethod(): StrategyType
	{
		return StrategyType.CertainSolutions;
	}

}

/**
 *

 else if (bestCandidates.size() == 1 || certainClearFound ) { // if there is only one solution or the solutions are certainties

        		// register all the moves
        		for (CandidateLocation cl: bestCandidates) {
        			Action move = cl.buildAction(MoveMethod.PROBABILITY_ENGINE);
            		//if (move.getAction() == Action.FLAG) {
            		//	boardState.setFlagConfirmed(move);
            		//}
        	    	// let the boardState decide what to do with this action
        			boardState.setAction(move);

        		}

        		// if we have a certain clear then also register all the mines
        		if (certainClearFound) {
            		display("Found " + pe.getMines().size() + " mines using the probability engine");
            		for (Location loc: pe.getMines()) {
            	    	// let the boardState decide what to do with this action
            			boardState.setAction(new Action(loc, Action.FLAG, MoveMethod.PROBABILITY_ENGINE, "",  BigDecimal.ONE));
            		}

        		}

    	        Action[] moves = boardState.getActions().toArray(new Action[0]);
    			fm = new FinalMoves(moves);

    		}
 */