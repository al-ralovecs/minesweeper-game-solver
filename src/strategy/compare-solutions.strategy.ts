import {AbstractStrategy, StrategyType} from "./abstract-strategy";
import ProbabilityDistributionDto, {PROBABILITY_ENGINE_TOLERANCE} from "../dto/probability-distribution.dto";
import WitnessWebDto from "../dto/witness-web.dto";
import BoardStateDto from "../dto/board-state.dto";
import ProbabilityEngineService from "../service/probability-engine.service";

export default class CompareSolutionsStrategy extends AbstractStrategy
{
    private readonly wholeEdge: WitnessWebDto;
    private readonly probabilityDistribution: ProbabilityDistributionDto;

    public constructor(
        boardState: BoardStateDto,
        wholeEdge: WitnessWebDto,
        probabilityDistribution: ProbabilityDistributionDto
    ) {
        super(boardState);

        this.wholeEdge = wholeEdge;
        this.probabilityDistribution = probabilityDistribution;
    }

    protected get isStrategyApplicable(): boolean
    {
        return 1 < ProbabilityEngineService.getBestCandidates(
                this.boardState,
                this.probabilityDistribution,
                PROBABILITY_ENGINE_TOLERANCE
            ).length
    }

    protected applyStrategy(): void
    {
    }

    protected get getMoveMethod(): StrategyType
    {
        return StrategyType.CompareRemainingSolutions;
    }

}

/**
 // fetch the best candidates from the edge
 List<CandidateLocation> bestCandidates = pe.getBestCandidates(PROB_ENGINE_TOLERENCE);
 *
 {    // evaluate which of the best candidates to choose
        		display("About to evaluate best candidates -->");
        		evaluateLocations.evaluateLocations(bestCandidates);
        		display("<-- Done");

        		evaluateLocations.showResults();

        		Action[] moves = evaluateLocations.bestMove();
        		fm = new FinalMoves(moves);

    		}
 */