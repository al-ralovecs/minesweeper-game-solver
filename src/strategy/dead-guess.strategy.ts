import {AbstractStrategy, StrategyType} from './abstract-strategy';

import BoardStateDto from '../dto/board-state.dto';
import WitnessWebDto from '../dto/witness-web.dto';
import ProbabilityDistributionDto from '../dto/probability-distribution.dto';

import AreaDto from '../dto/area.dto';
import LocationDto from '../dto/location.dto';

import guess from './guess-methods/chose-guess.method';

export default class DeadGuessStrategy extends AbstractStrategy
{
    private readonly wholeEdge: WitnessWebDto;
    private readonly probabilityDistribution: ProbabilityDistributionDto;

    private deadLocations: AreaDto;

    public constructor(
        boardState: BoardStateDto,
        wholeEdge: WitnessWebDto,
        probabilityDistribution: ProbabilityDistributionDto,
        deadLocations: AreaDto
    ) {
        super(boardState);

        this.wholeEdge = wholeEdge;
        this.probabilityDistribution = probabilityDistribution;

        this.deadLocations = deadLocations;
    }

    // if all the locations are dead
    protected get isStrategyApplicable(): boolean
    {
        const allWitnesses: LocationDto[] = this.boardState.getAllLivingWitnesses;
        const allWitnessedSquares: AreaDto = this.boardState.getUnrevealedArea(allWitnesses);

        return 0 !== this.deadLocations.size
            && this.deadLocations.size === allWitnessedSquares.size;
    }

    // then just use any one
    protected applyStrategy(): void
    {
        guess(this.boardState, this.wholeEdge, this.probabilityDistribution, this.getMoveMethod);
    }

    protected get getMoveMethod(): StrategyType
    {
        return StrategyType.DeadGuess;
    }
}
