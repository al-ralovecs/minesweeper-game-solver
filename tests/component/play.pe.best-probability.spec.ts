import {disposition} from '../__fixtures__/state/16x16x40.pe.best-probability.json';
import BoardDto from '../../src/dto/board.dto';
import Play from '../../src/component/play';
import Binomial from '../../src/utility/binomial';
import ActionDto, {ActionType} from "../../src/dto/action.dto";
import LocationDto from "../../src/dto/location.dto";
import {StrategyType} from "../../src/strategy/abstract-strategy";

describe('Play: Probability Engine results', () => {
    test('check if probability distribution matches', () => {
        const board = new BoardDto(disposition);
        const binomialEngine: Binomial = new Binomial(1000000, 100);
        const play: Play = new Play(board, binomialEngine, 40);

        expect(play.getNextMove).toStrictEqual(
            new ActionDto(new LocationDto(2, 7), ActionType.Clear, StrategyType.CompareRemainingSolutions, 0.925609)
        );

        expect(play.getProbabilityDistribution.offEdgeProbability).toBe(0.87487);
        expect(play.getProbabilityDistribution.cutOffProbability).toBe(0.88858464);
        expect(play.getProbabilityDistribution.bestProbability).toBe(0.925609);

        expect(play.getProbabilityDistribution.finalSolutionsCount).toBe(8007791515419721373659802075269356n);

        expect(play.getProbabilityDistribution.boxProb).toMatchSnapshot();
    });
});
