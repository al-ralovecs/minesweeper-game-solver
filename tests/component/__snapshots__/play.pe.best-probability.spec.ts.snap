// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play: Probability Engine results check if probability distribution matches 1`] = `
Array [
  0.517883,
  0.48211800000000005,
  0.925609,
  0.925609,
  0.666667,
  0.40772600000000003,
  0.925609,
  0.37053,
  0.258942,
  0.37053,
  0.37053,
  0.925609,
  0.543157,
]
`;
